lib/src/Pos/Util/UserPublic.hs

```diff
@@ -0,0 +1,287 @@
+import           Serokell.Util.Text (listJson)
+import           System.Directory (doesFileExist)
+import           System.Directory (renameFile)
+import           System.FileLock (FileLock, SharedExclusive (..), lockFile, unlockFile,
+                                  withFileLock)
+import           System.FilePath (takeDirectory, takeFileName)
+import           System.IO (hClose, openBinaryTempFile)
+import           System.Wlog (WithLogger)
+import           Test.QuickCheck (Arbitrary (..))
+import           Test.QuickCheck.Arbitrary.Generic (genericArbitrary, genericShrink)
+import           Universum
+
+import           Pos.Binary.Class (Bi (..), Cons (..), Field (..), decodeFull', deriveSimpleBi,
+                                   encodeListLen, enforceSize, serialize')
+import           Pos.Binary.Crypto ()
+import           Pos.Core (Address, addressF, makeRootPubKeyAddress)
+import           Pos.Crypto (PublicKey, PublicKey)
+
+import           Test.Pos.Crypto.Arbitrary ()
+
+#ifdef POSIX
+import           Formatting (oct, sformat)
+import qualified System.Posix.Files as PSX
+import qualified System.Posix.Types as PSX (FileMode)
+import           System.Wlog (logWarning)
+#endif
+
+-- Because of the Formatting import
+{-# ANN module ("HLint: ignore Use fewer imports" :: Text) #-}
+
+-- | Describes HD wallets keyfile content
+data WalletUserPublic = WalletUserPublic
+    { _wupKey        :: PublicKey           -- ^ root key of wallet set
+    , _wupWalletName :: Text                -- ^ name of wallet
+    , _wupAccounts   :: [(Word32, Text)]    -- ^ accounts coordinates and names
+    , _wupAddrs      :: [(Word32, Word32)]  -- ^ addresses coordinates
+    } deriving (Show, Generic)
+
+deriving instance Eq PublicKey => Eq WalletUserPublic
+
+instance Arbitrary WalletUserPublic where
+    arbitrary = genericArbitrary
+    shrink = genericShrink
+
+makeLenses ''WalletUserPublic
+
+instance Buildable WalletUserPublic where
+    build WalletUserPublic{..} =
+        bprint ("{ root address = "%addressF%", name = "%build%
+                ", accounts = "%pairsF%", addresses = "%pairsF%" }")
+        (makeRootPubKeyAddress _wupKey)
+        _wupWalletName
+        _wupAccounts
+        _wupAddrs
+      where
+        pairsF :: (Buildable a, Buildable b) => Format r ([(a, b)] -> r)
+        pairsF = later $ mconcat . map (uncurry $ bprint ("("%build%", "%build%")"))
+
+deriveSimpleBi ''WalletUserPublic [
+    Cons 'WalletUserPublic [
+        Field [| _wupKey        :: PublicKey          |],
+        Field [| _wupWalletName :: Text               |],
+        Field [| _wupAccounts   :: [(Word32, Text)]   |],
+        Field [| _wupAddrs      :: [(Word32, Word32)] |]
+    ]]
+
+-- | User public data.
+data UserPublic = UserPublic
+    { _upKeys    :: [PublicKey]
+    , _upWallet  :: Maybe WalletUserPublic
+    , _upPath    :: FilePath
+    , _upLock    :: Maybe FileLock
+    } deriving (Generic)
+
+deriving instance Eq PublicKey => Eq UserPublic
```

Comment:
This kind of constraint -- a fully concrete constraint -- is a big antipattern. It means that either:

1. This constraint is unsatisfiable, and errors will be deferred to the use site, with *tremendously* bad error messages
2. This constraint relies on an orphan instance that isn't in scope (bring it in scope, ideally by defining it next to the type in question)

These constraints cause big problems down the line, and they only *defer* the error messages. Please avoid using them.

As it happens, the type `PublicKey` already has an `Eq` instance, so this constraint is unnecessary.